### DISCUSSION TITLE
Implementing a dynamic "dict of metrics" efficiently

### DISCUSSION THREAD

**Discussion by orweiser-code** [Lightning-AI/torchmetrics#1834] [Ideas]
specificaly - apply one metric (e.g. MeanMetric) on a dict of values, where the dict keys are not known in advance (i.e. should add elements depending on the input dict of values to the update / forward functions.

simply creating a class that holds a dict of MeanMetric objects is very inefficient in a ddp setting.

implementing a metric that changes states shape dynamically seems unsupported.

what is your suggested approach, and do you intend to implement something like this soon?

**Comment by Borda** [comment]
Thanks for the question about applying a metric like `MeanMetric` dynamically to a dict with unknown keys. As you noted, creating a dict of separate metric objects is inefficient in DDP, and dynamic state shapes aren’t currently well supported in `torchmetrics`.
If possible, pre-define keys or create a custom metric managing dynamic states manually. Also, check the latest `torchmetrics` releases and discussions—this feature may have improved recently.
If it’s important for your use case, consider opening a feature request or discussing it with the maintainers on GitHub.
